### ISSUE TITLE
PGroonga and truncate table: too many files

### ISSUE THREAD

**Issue by JasonTrue** [pgroonga/pgroonga#72]
I've got a rails app and a bunch of tests that rely on a Database Cleaner task. Unfortunately, when I run tests, I get a lot of errors that, on my mac, look like this:

```
PG::InvalidName: ERROR:  pgroonga: object isn't found: <Sources584751>
```

And:

```
            ActiveRecord::StatementInvalid:
              PG::SystemError: ERROR:  pgroonga: failed to create table: <Lexicon589505_0>: system call error: Too many open files: failed to open file info path: <base/582944/pgrn.0000221>
```
(This one occurs when the database cleaner is invoking truncate table).

In our Travis CI build, I get other errors that are a bit less consistent (something about connecting to the psql socket, or unexpected model errors).

It looks like this only reproduces after I've created a pgroonga index. Merely enabling the extension does not reproduce the issue.

I've tried this with both single-column and multi-column indices and both seem to have this problem. Is there a workaround for this that I might be able to use? Obviously truncating tables isn't a very common use case in production but if I can't get tests to cooperate I won't be able to move it to production.

**Comment by JasonTrue** [comment]
Unfortunately, I don't think this will be a good solution. And the errors are even less predictable on travis, which is running on ubuntu. 

Here are 2 examples of error that occurred under Travis/Ubuntu when we had pgroonga indexes set up:

```
 ActiveRecord::StatementInvalid:
       PG::InternalError: ERROR:  negative bitmapset member not allowed
```
    
```
ActiveRecord::StatementInvalid:
       PG::InternalError: ERROR:  snapshot reference 0x7f5de80b4010 is not owned by resource owner Portal
```

They don't really occur in the same spot between test runs so I'm not sure of root cause.

Is there a way I can check how many files are being opened?

**Comment by kou** [comment]
`ls /proc/${PID}/fd/ | wc -l`

**Comment by JasonTrue** [comment]
BSD/OSX doesn't have quite the same option, so I tried to use lsof and activity monitor and I'm not much more enlightened. However, 
I raised the files per process limit on my macos machine from 24567 to 40000 and this does not seem to help.
```sudo sysctl -w kern.maxfilesperproc=40000```

On the first failure, I get a spurious:

ActiveRecord::StatementInvalid:
       PG::InvalidParameterValue: ERROR:  pgroonga: failed to set column value: [ii][column][update] record ID is nil

Then the same too many file handles all over again.

I'm a little confused about things like this, as well:
```                PG::SystemError: ERROR:  pgroonga: failed to create table: <Lexicon662763_0>: system call error: Too many open files: failed to open file info path: <base/628602/pgrn.000022D>```

The rather wide variety of errors suggests to me that something besides the file handle limit is going on.

**Comment by JasonTrue** [comment]
For what it's worth, ```lsof -c postgres |wc -l``` returns around 130 when postgres is nearly idle, stays in the hundreds for the first 300 tests or so, then quickly spikes up to 1152. Once the number is over 1000 or so, tests start to fail on the mac.

**Comment by JasonTrue** [comment]
No, the configuration setting is pretty explicitly maxfilesperproc. The system limit is 49152, via ```sysctl kern.maxfiles```. Either one of those limits is dramatically more than appear to be in use even when errors start ahapping.

I may be able to get Vacuum to do the trick. I've tried changing the strategy for cleaning to use transactions instead of truncation but that may be causing a different class of errors.

**Comment by JasonTrue** [comment]
I found another way to increase the limit (configuring launchctl). However, the numbers keep getting more extreme (11000+ file descriptors open based on lsof)  and the sort of random errors we experienced on Travis are now happening.

```
      ActiveRecord::StatementInvalid:
        PG::InvalidParameterValue: ERROR:  pgroonga: failed to set column value: [ii][column][update] record ID is nil
```

Here are some lines from pgroonga_log (I've minimized our index to only use one column instead of the multicolumn index I actually need.

```
2018-07-18 11:56:15.950805|n|4984: DDL:532:table_create BuildingSources841630
2018-07-18 11:56:15.951167|n|4984: spec:532:update:BuildingSources841630:48(table:hash_key):0
2018-07-18 11:56:15.951288|n|4984: DDL:526:table_create Lexicon841630_0
2018-07-18 11:56:15.951681|n|4984: spec:526:update:Lexicon841630_0:49(table:pat_key):0
2018-07-18 11:56:15.951768|n|4984: DDL:524:column_create BuildingSources841630 first_name_alphabet
2018-07-18 11:56:15.952095|n|4984: spec:524:update:BuildingSources841630.first_name_alphabet:64(column:fix_size):526(Lexicon841630_0)
2018-07-18 11:56:15.952163|n|4984: DDL:527:column_create Lexicon841630_0 index
2018-07-18 11:56:15.953092|n|4984: spec:527:update:Lexicon841630_0.index:72(column:index):532(BuildingSources841630)
2018-07-18 11:56:15.953195|n|4984: DDL:527:set_source Lexicon841630_0.index BuildingSources841630.first_name_alphabet
2018-07-18 11:56:15.953262|n|4984: spec:524:update:BuildingSources841630.first_name_alphabet:64(column:fix_size):526(Lexicon841630_0)
2018-07-18 11:56:15.953330|n|4984: spec:527:update:Lexicon841630_0.index:72(column:index):532(BuildingSources841630)
```

**Comment by kou** [comment]
Thanks for providing pgroonga.log. Can you upload full pgroonga.log? It should have log on `pgroonga: failed to set column value: [ii][column][update] record ID is nil`.

Can you create a sample Rails application that reproduces this case? It's difficult to look into this case without reproducing this case on local.

**Comment by JasonTrue** [comment]
[justyesterday.log](https://github.com/pgroonga/pgroonga/files/2208315/justyesterday.log)

I've attached slightly reduced log (just the logs from 7-18). I'm not sure how to reproduce this issue in a minimalist rails app because it appears to happen only after a few hundred test cases that do Truncate Table between tests have run. I can try to do that when I have time, but I'm not sure when I'll get to it, as my schedule constraints have led me to use an alternate indexing solution for now.

**Comment by kou** [comment]
Thanks. I found that Groonga doesn't log enough information. I've added it to master.

I came up with an idea that solves this case: PGroonga closes old PGroonga indexes when an PGroonga index is recreated (`TRUNCATE` causes this). If PGroonga closes old PGroonga indexes, `postgresql` process can decrease opened file descriptors.

**Comment by kou** [comment]
I've implemented the idea on mater. You can try it by setting `PGROONGA_MASTER=yes` environment variable on Travis CI.

**Comment by kou** [comment]
I've released 2.1.1 with this feature.
